### DISCUSSION TITLE
AllSky environment variables from inside a custom shell script

### DISCUSSION THREAD

**Discussion by laheller** [AllskyTeam/allsky-modules#178] [Q&A] [answered]
Hi @Alex-developer 

I would like to execute my custom shell script after each nighttime image is taken. According to the documentation there is [allsky_script](https://github.com/AllskyTeam/allsky-modules/tree/master/allsky_script) module for this purpose. I would like to ask, whether is it possible to access **AllSky enviroment variables** from within my custom script. Actually, in the beginning I only need is the environment variable which tells the full path to the fresh image from the camera. Is it possible to access those variables from a bash script?

BR,

Ladislav

**Comment by laheller** [comment, marked as answer]
Meanwhile I found the answer to my own question. All the below variables are accessible from inside the custom bash script executed by the **allsky_script** module (real example, got it by simply calling `set > output1.txt` from custom script):

> ALLSKY_ABORTEDSAVEIMAGE=saveImage.txt
> ALLSKY_ABORTEDTIMELAPSE=timelapse.txt
> ALLSKY_ABORTEDUPLOADS=uploads.txt
> ALLSKY_ABORTS_DIR=/home/laheller/allsky/tmp/aborts
> ALLSKY_BAD_IMAGE_COUNT=/home/laheller/allsky/tmp/bad_image_count.txt
> ALLSKY_BIN=/home/laheller/allsky/bin
> ALLSKY_CONFIG=/home/laheller/allsky/config
> ALLSKY_DARKS=/home/laheller/allsky/darks
> ALLSKY_DEBUG_LEVEL=3
> ALLSKY_DOCUMENTATION=/home/laheller/allsky/html/documentation
> ALLSKY_EXTRA=/home/laheller/allsky/config/overlay/extra
> ALLSKY_FLOWTIMINGS=/home/laheller/allsky/tmp/flowtimings
> ALLSKY_FLOWTIMINGS_DAY=/home/laheller/allsky/tmp/flowtimings/day-average
> ALLSKY_FLOWTIMINGS_NIGHT=/home/laheller/allsky/tmp/flowtimings/night-average
> ALLSKY_GROUP=laheller
> ALLSKY_HOME=/home/laheller/allsky
> ALLSKY_IMAGES=/home/laheller/allsky/images
> ALLSKY_INSTALLATION_LOGS=/home/laheller/allsky/config/logs
> ALLSKY_INSTALL_DIR=allsky
> ALLSKY_LOG=/var/log/allsky.log
> ALLSKY_MESSAGES=/home/laheller/allsky/config/messages.txt
> ALLSKY_MODULES=/home/laheller/allsky/config/modules
> ALLSKY_MODULE_LOCATION=/opt/allsky
> ALLSKY_NOTIFICATION_IMAGES=/home/laheller/allsky/notification_images
> ALLSKY_NOTIFICATION_LOG=/home/laheller/allsky/tmp/notification_log.txt
> ALLSKY_OVERLAY=/home/laheller/allsky/config/overlay
> ALLSKY_OWNER=laheller
> ALLSKY_PERIODIC_LOG=/var/log/allskyperiodic.log
> ALLSKY_PYTHON_VENV=/home/laheller/allsky/venv
> ALLSKY_REBOOT_NEEDED=/home/laheller/allsky/tmp/reboot_needed.txt
> ALLSKY_REMOTE_WEBSITE_CONFIGURATION_FILE=/home/laheller/allsky/config/remote_configuration.json
> ALLSKY_REMOTE_WEBSITE_CONFIGURATION_NAME=remote_configuration.json
> ALLSKY_REPO=/home/laheller/allsky/config_repo
> ALLSKY_SCRIPTS=/home/laheller/allsky/scripts
> ALLSKY_TIMELAPSE_PID_FILE=/home/laheller/allsky/tmp/timelapse-pid.txt
> ALLSKY_TMP=/home/laheller/allsky/tmp
> ALLSKY_VARIABLE_SET=true
> ALLSKY_VERSION=v2023.05.01_04
> ALLSKY_VERSION_FILE=/home/laheller/allsky/version
> ALLSKY_WEBSITE=/home/laheller/allsky/html/allsky
> ALLSKY_WEBSITE_BRANCH_FILE=/home/laheller/allsky/html/allsky/branch
> ALLSKY_WEBSITE_CONFIGURATION_FILE=/home/laheller/allsky/html/allsky/configuration.json
> ALLSKY_WEBSITE_CONFIGURATION_NAME=configuration.json
> ALLSKY_WEBSITE_VERSION_FILE=/home/laheller/allsky/html/allsky/version
> ALLSKY_WEBSITE_VIEWSETTINGS_DIRECTORY=/home/laheller/allsky/html/allsky/viewSettings
> ALLSKY_WEBSITE_VIEWSETTINGS_DIRECTORY_NAME=viewSettings
> ALLSKY_WEBUI=/home/laheller/allsky/html
> ARGS_FILE=/home/laheller/allsky/tmp/capture_args.txt
> AS_AUTOEXPOSURE=0
> AS_AUTOGAIN=0
> AS_AUTOUSB=1
> AS_AUTOWB=0
> AS_BIN=1
> AS_BIT_DEPTH=8
> AS_BRIGHTNESS=100
> AS_CAMERA_MODEL=ASI120MC-S
> AS_CAMERA_TYPE=ZWO
> AS_DARKFRAME=0
> AS_DATE=20241112
> AS_EXPOSURE_US=30000000
> AS_FLIP=none
> AS_FULLMEAN=245
> AS_GAIN=50
> AS_LH_CPU_TEMP='23.8 °C'
> AS_MEAN=255
> AS_TEMPERATURE_C=8
> AS_TEMPERATURE_F=46
> AS_TIME=220609
> AS_USB=50
> AS_WBB=95
> AS_WBR=52
> AS_eOVERLAY=1
> AS_sAUTOAWB=
> AS_sAUTOEXPOSURE=
> AS_sAUTOGAIN=
> AS_sEXPOSURE='30.0 sec'
> AUTO_STRETCH=true
> AUTO_STRETCH_AMOUNT=5
> AUTO_STRETCH_MID_POINT=15%
> BASH=/bin/bash
> BASHOPTS=checkwinsize:cmdhist:complete_fullquote:extquote:force_fignore:globasciiranges:globskipdots:hostcomplete:interactive_comments:patsub_replacement:progcomp:promptvars:sourcepath
> BASH_ALIASES=()
> BASH_ARGC=()
> BASH_ARGV=()
> BASH_CMDS=()
> BASH_LINENO=([0]="0")
> BASH_LOADABLES_PATH=/usr/local/lib/bash:/usr/lib/bash:/opt/local/lib/bash:/usr/pkg/lib/bash:/opt/pkg/lib/bash:.
> BASH_SOURCE=([0]="/home/laheller/test1.sh")
> BASH_VERSINFO=([0]="5" [1]="2" [2]="15" [3]="1" [4]="release" [5]="aarch64-unknown-linux-gnu")
> BASH_VERSION='5.2.15(1)-release'
> BRIGHTNESS_THRESHOLD=0.2
> CAMERA_MODEL=ASI120MC-S
> CAMERA_TYPE=ZWO
> CAPTURE=capture_ZWO
> CAPTURE_SAVE_DIR=/home/laheller/allsky/tmp
> CC_FILE=/home/laheller/allsky/config/cc.json
> CONFIG_SH_VERSION=1
> CROP_HEIGHT=480
> CROP_IMAGE=false
> CROP_OFFSET_X=0
> CROP_OFFSET_Y=0
> CROP_WIDTH=640
> CURRENT_IMAGE=/home/laheller/allsky/tmp/image-20241112220609.jpg
> DATE_NAME=20241112
> DAYS_TO_KEEP=14
> DAY_OR_NIGHT=NIGHT
> DEV=
> DIRSTACK=()
> EUID=1000
> EXIT_ERROR_STOP=100
> EXIT_NO_CAMERA=101
> EXIT_OK=0
> EXIT_RESET_USB=99
> EXIT_RESTARTING=98
> EXTENSION=jpg
> FFLOG=warning
> FILENAME=image
> FPS=25
> FREQUENCY_FILE=/home/laheller/allsky/tmp/IMG_UPLOAD_FREQUENCY.txt
> FULL_FILENAME=image.jpg
> GITHUB_ALLSKY_PACKAGE=allsky
> GITHUB_MAIN_BRANCH=master
> GITHUB_RAW_ROOT=https://raw.githubusercontent.com/AllskyTeam
> GITHUB_ROOT=https://github.com/AllskyTeam
> GITHUB_WEBSITE_PACKAGE=allsky-website
> GREEN=
> GROUPS=()
> HOME=/home/laheller
> HOSTNAME=allsky-ladsi
> HOSTTYPE=aarch64
> IFS=$' \t\n'
> IMG_CREATE_THUMBNAILS=false
> IMG_DIR=current/tmp
> IMG_HEIGHT=1520
> IMG_RESIZE=false
> IMG_UPLOAD=true
> IMG_UPLOAD_FREQUENCY=5
> IMG_UPLOAD_ORIGINAL_NAME=false
> IMG_WIDTH=2028
> INVOCATION_ID=2b7d98fd2cdc4baaa70ee2bdf338719d
> JOURNAL_STREAM=8:15161
> KEEP_SEQUENCE=false
> KEOGRAM=true
> KEOGRAM_EXTRA_PARAMETERS='--font-size 1.0 --font-line 1 --font-color '\''255 255 255'\'''
> LANG=C.UTF-8
> LANGUAGE=C.UTF-8
> LAST_CHANGED='2024-11-12 21:02:20'
> LD_LIBRARY_PATH=/home/laheller/allsky/venv/lib/python3.11/site-packages/cv2/../../lib64:
> LOCALE=C.UTF-8
> LOGNAME=laheller
> MACHTYPE=aarch64-unknown-linux-gnu
> ME2=darkSubtract.sh
> MSG=
> NC=
> NEEDS_REBOOT=false
> NO_LINK_=3
> NUM=1
> OLDPWD=/
> OLD_WEBSITE_LOCATION=/var/www/html/allsky
> OLD_WEBUI_LOCATION=/var/www/html
> ON_TTY=0
> OPTERR=1
> OPTIND=1
> OPTIONS_FILE=/home/laheller/allsky/config/options.json
> OSTYPE=linux-gnu
> PATH=/home/laheller/allsky/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> PIX_FMT=yuv420p
> PI_OS=bookworm
> POST_INSTALLATION_ACTIONS=/home/laheller/allsky/config/logs/post-installation_actions.txt
> PPID=3930
> PRIOR_ALLSKY_DIR=/home/laheller/allsky-OLD
> PS4='+ '
> PWD=/home/laheller/allsky
> PYTHON_VENV_ACTIVATED=true
> QT_QPA_FONTDIR=/home/laheller/allsky/venv/lib/python3.11/site-packages/cv2/qt/fonts
> QT_QPA_PLATFORM_PLUGIN_PATH=/home/laheller/allsky/venv/lib/python3.11/site-packages/cv2/qt/plugins
> RED=
> REMOVE_BAD_IMAGES=false
> REMOVE_BAD_IMAGES_THRESHOLD_HIGH=90
> REMOVE_BAD_IMAGES_THRESHOLD_LOW=1
> REPO_WEBCONFIG_FILE=/home/laheller/allsky/config_repo/configuration.json.repo
> RESETTING_USB_LOG=/home/laheller/allsky/tmp/resetting_USB.txt
> RESIZE_UPLOADS=false
> RESIZE_UPLOADS_HEIGHT=720
> RESIZE_UPLOADS_WIDTH=962
> RPi_COMMAND_TO_USE=
> SEE_LOG_MSG='See /var/log/allsky.log'
> SETTINGS_FILE=/home/laheller/allsky/config/settings.json
> SHELL=/bin/bash
> SHELLOPTS=braceexpand:hashall:interactive-comments
> SHLVL=3
> STARTRAILS=true
> STARTRAILS_EXTRA_PARAMETERS=
> SYSTEMD_EXEC_PID=3743
> TAB=$'\t'
> TEMP=/home/laheller/allsky/tmp/ZWO_cameras.txt
> TERM=dumb
> THUMBNAIL_SIZE_X=100
> THUMBNAIL_SIZE_Y=75
> TIMELAPSE=true
> TIMELAPSEHEIGHT=0
> TIMELAPSEWIDTH=0
> TIMELAPSE_BITRATE=5000k
> TIMELAPSE_EXTRA_PARAMETERS=
> TIMELAPSE_MINI_BITRATE=2000k
> TIMELAPSE_MINI_FORCE_CREATION=false
> TIMELAPSE_MINI_FPS=5
> TIMELAPSE_MINI_FREQUENCY=5
> TIMELAPSE_MINI_HEIGHT=600
> TIMELAPSE_MINI_IMAGES=20
> TIMELAPSE_MINI_UPLOAD_THUMBNAIL=true
> TIMELAPSE_MINI_UPLOAD_VIDEO=true
> TIMELAPSE_MINI_WIDTH=800
> TIMELAPSE_UPLOAD_THUMBNAIL=true
> UHUBCTL_PATH=
> UHUBCTL_PORT=2
> UID=1000
> UPLOAD_KEOGRAM=true
> UPLOAD_STARTRAILS=true
> UPLOAD_VIDEO=true
> USER=laheller
> USE_NOTIFICATION_IMAGES=1
> VCODEC=libx264
> VIRTUAL_ENV=/home/laheller/allsky/venv
> VIRTUAL_ENV_PROMPT='(venv) '
> WEBSERVER_GROUP=www-data
> WEBUI_DATA_FILES=/home/laheller/allsky_info.txt
> WEB_DAYS_TO_KEEP=0
> YELLOW=
> ZWOdev=/dev/bus/usb/002/002
> _=/home/laheller/allsky/venv/bin/python3
> _OLD_VIRTUAL_PS1=
> cDEBUG=
> wBOLD='<b>'
> wBR='<br>'
> wDEBUG='<span style='\''color: #FF9800'\''>'
> wERROR='<span style='\''color: red'\''>'
> wNBOLD='</b>'
> wNC='</span>'
> wOK='<span style='\''color: green'\''>'
> wWARNING='<span style='\''color: #FF9800'\''>'
>

**Comment by Alex-developer** [comment]
Sorry for the late reply but it looks like you have found the answer :-)

**Reply by laheller** [reply]
@Alex-developer

Sorry for "reopening" the topic, but now I have a specific question to those environment variables:
Which one to take when I want to process the current captured image and let that processed image also show on the allsky website's live view? I tried the one referenced by the environment variable
`CURRENT_IMAGE`,
so if the value was for example
`$HOME/allsky/tmp/image-20241112220609.jpg`,
I changed that image. But when I checked it on the allsky website's live view, I did not see any change.

**Reply by Alex-developer** [reply]
No problem reopening :-)

I think the problem you are having is the 'flow manager', the code that runs the modules, is overwriting your change.

The 'flow manager' first runs the loadImage module which loads the CURRENT_IMAGE into a numpy array in the shared module, this can be access from within a  module as s.image. At the end of the 'flow manager' the saveimage module is called which saves the image and I suspect is overwriting any changes you made to it. The reason I do this is to keep the image in memory between module runs.

What does your script do? Probably the best option is to write a simple module to do the same

Happy to help with a module if its needed

Alex

**Reply by laheller** [reply]
Hi
@Alex-developer 

My script does the following:
- Takes the _$CURRENT_IMAGE_ and using _ffmpeg_ crops it to keep only the central area (final cropped image is one quarter of the original one).
- Using [astrometry.net](https://astrometry.net/) plate solver (a prerequisite, can be installed via package manager) it detects the stars on the cropped image and it draws the constellation lines on the cropped image as an overlay.
-  Finally again using _ffmpeg_ it overlays the cropped image made up by constellation lines over the original image exactly at the central area.

Notes:
The reason why it crops before the actual plate solving is that the plate solver cannot deal with the full field image taken by a fisheye lens which distorts the field around the central part.

An example of the processed image:
![image](https://github.com/user-attachments/assets/9e6e4c7d-714c-4479-96fd-6677ca25f6ee)

Anyway, the script works perfectly and does exactly what I want, the only problem is as I wrote, that the processed image does not appear on the allsky website's live view.

So if I understand correctly, such a thing is not possible to do from a shell script and I need to create a custom Python module, right?

BR,

Ladislav

**Reply by Alex-developer** [reply]
I was thinking about this and there might be another way. 

When a script is run and completes it can send a specific exit code back to the script module that would then trigger it to reload the image. That might be easier for you and others.

So for example at the end of your script you would add something like

exit 27

Which would indicate that the image needs reloading from disk again.

There are some caveats to this approach the biggest being that you would have to run the script module before any other modules amend the image,.

I'll have a play around later and see what I can do.

Out of interest how long is the plate solving taking, when I have used astrometry.net its been very slow

**Reply by laheller** [reply]
@Alex-developer 

Thanks, I'll try the trick using the exit code 27.

> Out of interest how long is the plate solving taking, when I have used astrometry.net its been very slow

It's all about the correct input parameters. Means if we provide the good estimate FOV size (range) to the solver, it's just about about **1.5 seconds**, like in the below real example. I used the _time_ command to measure the execution time:

> time **solve-field** -O -l 300 -D $HOME/solve -L 40 -H 70 -u dw -z 1 -d 20 --crpix-center -P $HOME/solve/img_crop1.ppm -p -9 -y --uniformize 0 $HOME/img_crop1.jpg
>
> Reading input file 1 of 1: "/home/laheller/img_crop1.jpg"...
> jpegtopnm: WRITING PPM FILE
> Read file stdin: 640 x 480 pixels x 1 color(s); maxval 255
> Using 8-bit output
> Extracting sources...
> simplexy: found 260 sources.
> Solving...
> Reading file "/home/laheller/solve/img_crop1.axy"...
>   log-odds ratio 39.6337 (1.63188e+17), 34 match, 0 conflict, 88 distractors, 76 index.
>   RA,Dec = (34.1192,46.9598), pixel scale 361.196 arcsec/pix.
>   Hit/miss:   Hit/miss: +---------+-----+-----+--+--+-------+--------++--+----+---+-+--+-----+--+--+--------++-+-----+---+-+
> Field 1: solved with index index-tycho2-19.littleendian.fits.
> Field 1 solved: writing to file /home/laheller/solve/img_crop1.solved to indicate this.
> Field: /home/laheller/img_crop1.jpg
> Field center: (RA,Dec) = (33.519557, 46.807122) deg.
> Field center: (RA H:M:S, Dec D:M:S) = (02:14:04.694, +46:48:25.639).
> Field size: 60.389 x 46.4388 degrees
> Field rotation angle: up is 151.84 degrees E of N
> Field parity: neg
> Creating new FITS file "/home/laheller/solve/img_crop1.new"...
> 
> 
> real    0m1.425s
> user    0m1.310s
> sys     0m0.128s
> 

The most important parameters for the [solve-field](https://manpages.debian.org/testing/astrometry.net/solve-field.1.en.html) (part of the astrometry.net suite) command:

- -l 300: Maximum allowed time in seconds for the plate solver
- -L 40: The lower bound of the estimated FOV size in degrees
- -H 70: The upped bound of the estimated FOV size in degrees
- -u dw: Whether the above two parameter values are in degrees
- -z 1: Downsample factor of the input image, this time 1x, means no downsamle

Also we can reduce the execution time by providing the input image to the solver in FITS format instead of JPG, because it only accepts FITS format. In case of JPG, PNG or any other format it tries to convert first to FITS.

I have an additional question to the original topic - image processing via a script:
How to do the processing ONLY on every N-th image?

Update:
Unfortunately adding exit 27 to the end of script did not help: I still don't see the changes on the allsky website.

BR,

Ladislav

**Reply by laheller** [reply]
@Alex-developer 

Based on the boilerplate example module I created a simple prototype module with the below code. It does what I want, however I would like to do the processing only on every N-th imge. Is that possible?

```python
import allsky_shared as s
import os
import subprocess
import cv2

metaData = {
    "name": "ladsi",
    "description": "ladsi module",
    "module": "allsky_ladsi",
    "events": [
        "night"
    ],
    "experimental": "true"
}

def ladsi(params, event):
    #s.log(1, 'Writing image to current.jpg')
    current_path = '{HOME}/current.jpg'.format(**os.environ)
    if os.path.exists(current_path): os.remove(current_path)

    cv2.imwrite(current_path, s.image)

    cropped_path = '{HOME}/cropped.jpg'.format(**os.environ)
    res = subprocess.run(['ffmpeg','-hide_banner','-i',f'{current_path}','-vf','crop=in_w/2:in_h/2,drawbox=width=in_w:height=in_h:color=red:thickness=1','-frames:v','1','-update','1','-y',f'{cropped_path}'])

    solve_path = '{HOME}/solve'.format(**os.environ)
    ppm_path = f'{solve_path}/current.ppm'
    res = subprocess.run(['solve-field','-O','-l','10','-D',f'{solve_path}','-L','40','-H','70','-u','dw','-z','1','-d','20','--crpix-center','-P',f'{ppm_path}','-p','-9','-y','--uniformize','0',f'{cropped_path}'])
    if (res.returncode != 0): return "Plate solve failed!"

    wcs_path = f'{solve_path}/cropped.wcs'
    tagged_path = f'{solve_path}/tagged.png'
    res = subprocess.run(['plot-constellations','-w',f'{wcs_path}','-i',f'{ppm_path}','-C','-o',f'{tagged_path}'])

    res = subprocess.run(['ffmpeg','-hide_banner','-i',f'{current_path}','-i',f'{tagged_path}','-filter_complex','overlay=main_w/4:main_h/4','-frames:v','1','-update','1','-y',f'{current_path}'])

    s.image = cv2.imread(f'{current_path}',cv2.IMREAD_COLOR)

    #s.log(1, 'Done!')
    return 'N/A'

```
BR,

Ladislav

**Reply by Alex-developer** [reply]
Yes there is :-)

allsky shared has a 'database' that you can save and read values from Have a look at the discord module as it implements this type of behaviour. Essentially it goes something like

If db entry doesnt exist set it to 0
Each time the module runs up the counter by one
if the counter exceeds your limit run your code and set the counter to zero

This will ONLY work in the day or night flows. If you want to do this in the periodic flow then its a slightly different approach

If its not obvious or you have any issues give me a shout

Cheers

Alex
